### PR TITLE
Log Wrapper for logstash implementation

### PR DESCRIPTION
**Pull request by AmarnatReddy**
Log Wrapper for logstash implementation

Log file : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-**9JI2QC**/
This is failed log but logstash entries are present

for all the log messages we have added metadata which contains run_id
with run_id we can filter all the logs corresponding to that run
URL : http://10.0.211.222:5601/app/logs/stream?flyoutOptions=(flyoutId:%273WI-3H0Bf41K1ck-IKoY%27,flyoutVisibility:hidden,surroundingLogsId:%273WI-3H0Bf41K1ck-IKoY%27)&logPosition=(end:now,position:(tiebreaker:14009,time:1640077573351),start:now-1d,streamLive:!f)

Now we have only changed in run.py and install_prereq.py files. Once we finalize on the implementation we can will go ahead and change in all the required places

**ELK Stack Server details :** 
http://10.0.211.222:5601/app/logs/stream?flyoutOptions=(flyoutId:%273WI-3H0Bf41K1ck-IKoY%27,flyoutVisibility:hidden,surroundingLogsId:%273WI-3H0Bf41K1ck-IKoY%27)&logPosition=(end:now,position:(tiebreaker:14009,time:1640077573351),start:now-1d,streamLive:!f)
username : elastic
password : changeme

Signed-off-by: Amarnath K <amk@amk.remote.csb>

# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarin Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: run.py [hunk 1]
```diff
@@ -374,7 +378,30 @@ def run(args):
     log_directory = args.get("--log-dir")
 
     run_id = generate_unique_id(length=6)
-    run_dir = create_run_dir(run_id, log_directory)
+    # run_dir = create_run_dir(run_id, log_directory)
+    c = TestMetaData(
+        run_id=run_id,
+        rhbuild=rhbuild,
+        rhcs="rhcs",
```
Comment (psathyan) [inline run.py:385]: Why do we have `rhcs=rhcs`? Should it be the version?

File: utility/log.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        handler = logstash.TCPLogstashHandler(
+            host=host,
+            port=port,
+            version=version,
+        )
+        handler.setLevel(self.log_level)
+        self._logger.addHandler(handler)
+
+        server = f"tcp://{host}:{port}"
+        self._logger.debug(f"Log events are also pushed to {server}")
```
Comment (psathyan) [inline utility/log.py:99]: Do we need this method?

File: utility/log.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        self.debug(f"Log events are also pushed to {_log_file}")
+
+    def close_file_handler(self):
+        """Close any file handler associated with the logger."""
+        handlers = self._logger.handlers[:]
+        for handler in handlers:
+            if isinstance(handler, logging.FileHandler):
+                handler.close()
+                self._logger.removeHandler(handler)
+
```
Comment (psathyan) [inline utility/log.py:165]: Do we need these methods?

File: utility/log.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        """
+        log = {
+            "info": self._logger.info,
+            "debug": self._logger.debug,
+            "warning": self._logger.warning,
+            "error": self._logger.error,
+            "exception": self._logger.exception,
+        }
+        extra = None
+        if self.config.get("logstash"):
```
Comment (psathyan) [inline utility/log.py:186]: Enable logstash in `run.py` to void an issue here.

**Review by psathyan — COMMENTED** [review]
Could you have a run with and without logstash enabled?

File: utility/log.py (hunk 3 above)
Comment (AmarnatReddy) [reply utility/log.py]: I have removed this methods

File: run.py (hunk 1 above)
Comment (AmarnatReddy) [reply run.py]: Changed rhcs to ceph Version

**Comment by AmarnatReddy** [comment]
> Could you have a run with and without logstash enabled?
Logs with out Logstash : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-21YHBJ/

Logs with Logstash : 
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-511ENP/
Logstash :http://10.0.211.222:5601/app/logs/stream?flyoutOptions=(flyoutId:HmWr4H0Bf41K1ck-KbsH,flyoutVisibility:hidden,surroundingLogsId:!n)&logPosition=(end:now,position:(tiebreaker:11,time:1640153362727),start:now-1d,streamLive:!t)&logFilter=(language:kuery,query:%27%22test_run_id%22:%22511ENP%22%27)

**Comment by AmarnatReddy** [comment]
Logs without logstash : https://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-TESBYP/
Logs with Logstash : 
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-L0KJAF/

http://10.0.211.222:5601/app/logs/stream?flyoutOptions=(flyoutId:WWan4X0Bf41K1ck--5hW,flyoutVisibility:hidden,surroundingLogsId:!n)&logPosition=(end:now,position:(tiebreaker:6,time:1640168421852),start:now-1d,streamLive:!f)&logFilter=(language:kuery,query:%27%22test_run_id%22:%22L0KJAF%22%27)

File: utility/log.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+    @property
+    def config(self) -> Dict:
+        """Return the CephCI run configuration."""
+        return TestMetaData()
+
+    @property
+    def run_id(self) -> str:
+        """Return the unique identifier of the execution run."""
+        return self.config["run_id"]
```
Comment (psathyan) [inline utility/log.py:61]: Opt to use `self.config.get("run_id")`... this should fix gate failures.

File: utility/log.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        extra.update(kwargs.get("metadata", {}))
+        log[level](message, *args, extra=extra, **kwargs)
+
+    def info(self, message: Any, *args, **kwargs) -> None:
+        """
+        Log with info level the provided message and extra data.
+
+        Args:
+            message (Any):      The message to be logged.
+            metadata (dict):    The metadata that would be sent along with the message.
```
Comment (psathyan) [inline utility/log.py:106]: Fix doc string... you would need to remove `metadata` from the below methods.

File: utility/log.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    """CephCI Logger object to help streamline logging."""
+
+    def __init__(self, filename="") -> None:
+        """
+        Initializes the logging mechanism based on the inputs provided."""
+        self._logger = logging.getLogger(filename)
+        self._log_level = logging.DEBUG
+        self._log_dir = None
+        self.log_format = LOG_FORMAT
+        self.module_name = None
```
Comment (psathyan) [inline utility/log.py:41]: You may want to add a method to check if logstash is enabled for writing the metadata...

Comment (AmarnatReddy) [reply utility/log.py:41]: This we are checking as part of TestMetadata creation in run.py

**Review by psathyan — COMMENTED** [review]
looks good to me

**Review by psathyan — APPROVED** [review]
Looks good to me

**Pull request merged**